File: mcp/src/tools/conclusions.ts
import { z } from "zod";

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { ToolContext } from "../types.js";
import { textResult, errorResult, workspaceIdSchema } from "../types.js";

export function register(server: McpServer, ctx: ToolContext) {
  // ── list_conclusions ────────────────────────────────────────────────
  server.registerTool(
    "list_conclusions",
    {
      description: [
        "List conclusions (facts and observations) that Honcho has derived about a peer (paginated).",
        "Use this to see what Honcho has learned. If no target is given, returns self-conclusions.",
        "Returns conclusion objects with pagination metadata.",
        "Pass include_deleted to also see retired conclusions and why each was removed.",
      ].join("\n"),
      inputSchema: {
        workspace_id: workspaceIdSchema(ctx),
        peer_id: z.string().describe("The observer peer."),
        target_peer_id: z
          .string()
          .optional()
          .describe(
            "Optional: list conclusions about this target. Omit for self-conclusions.",
          ),
        include_deleted: z
          .boolean()
          .optional()
          .describe(
            "Include retired conclusions, each carrying the reason it was removed.",
          ),
        page: z
          .number()
          .int()
          .min(1)
          .optional()
          .describe("Page number, 1-indexed. Retired conclusions past page one are only reachable this way."),
        size: z
          .number()
          .int()
          .min(1)
          .max(100)
          .optional()
          .describe("Results per page (default 50, max 100)."),
        filters: z
          .record(z.string(), z.unknown())
          .optional()
          .describe(
            'Additional filter criteria, e.g. {"level": "explicit"}. Merged with this peer pair.',
          ),
      },
    },
    async ({
      workspace_id,
      peer_id,
      target_peer_id,
      include_deleted,
      page: pageNumber,
      size,
      filters,
    }) => {
      try {
        const peer = await ctx.clientFor(workspace_id).peer(peer_id);
        const scope = target_peer_id
          ? peer.conclusionsOf(target_peer_id)
          : peer.conclusions;
        const page = await scope.list({
          includeDeleted: include_deleted,
          page: pageNumber,
          size,
          filters,
        });
        return textResult({
          conclusions: page.items.map((c) => ({
            id: c.id,
            content: c.content,
            observer_id: c.observerId,
            observed_id: c.observedId,
            session_id: c.sessionId,
            times_derived: c.timesDerived,
            created_at: c.createdAt,
            ...(c.deletedAt
              ? { deleted_at: c.deletedAt, removal: c.removal ?? null }
              : {}),
          })),
          total: page.total,
          page: page.page,
          pages: page.pages,
        });
      } catch (e) {
        return errorResult(
          `Failed to list conclusions: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── query_conclusions ───────────────────────────────────────────────
  server.registerTool(
    "query_conclusions",
    {
      description: [
        "Semantic search across a peer's conclusions.",
        "Use this to find specific knowledge Honcho has derived — more targeted than list_conclusions.",
        "Returns an array of matching conclusions ranked by relevance.",
      ].join("\n"),
      inputSchema: {
        workspace_id: workspaceIdSchema(ctx),
        peer_id: z.string().describe("The observer peer."),
        query: z.string().describe("Semantic search query."),
        target_peer_id: z
          .string()
          .optional()
          .describe("Optional: search conclusions about this target."),
        top_k: z
          .number()
          .optional()
          .describe("Max results to return."),
        filters: z
          .record(z.string(), z.unknown())
          .optional()
          .describe(
            'Optional: filter criteria, e.g. {"level": ["deductive", "inductive"]} to only return conclusions derived during dreaming. Levels: explicit (extracted directly from messages), deductive, inductive, contradiction. See https://honcho.dev/docs/v3/documentation/features/advanced/using-filters',
          ),
      },
    },
    async ({ workspace_id, peer_id, query, target_peer_id, top_k, filters }) => {
      try {
        const peer = await ctx.clientFor(workspace_id).peer(peer_id);
        const scope = target_peer_id
          ? peer.conclusionsOf(target_peer_id)
          : peer.conclusions;
        const conclusions = await scope.query(query, top_k, undefined, filters);
        return textResult(
          conclusions.map((c) => ({
            id: c.id,
            content: c.content,
            level: c.level,
            observer_id: c.observerId,
            observed_id: c.observedId,
            session_id: c.sessionId,
            times_derived: c.timesDerived,
            created_at: c.createdAt,
          })),
        );
      } catch (e) {
        return errorResult(
          `Query failed: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── create_conclusions ──────────────────────────────────────────────
  server.registerTool(
    "create_conclusions",
    {
      description: [
        "Manually create conclusions (facts/observations) about a peer.",
        "Use this to inject knowledge into Honcho that wasn't derived from conversation.",
        "Returns the number of conclusions created.",
      ].join("\n"),
      inputSchema: {
        workspace_id: workspaceIdSchema(ctx),
        peer_id: z.string().describe("The observer peer."),
        target_peer_id: z
          .string()
          .describe("The peer the conclusions are about."),
        conclusions: z.array(
          z.object({
            content: z.string(),
            action: z.enum(["create", "enrich"]),
            target_id: z.string().optional(),
            reason_for_entry: z.string(),
            search_query: z.string(),
            searched_conclusion_ids: z.array(z.string()),
            source_message_ids: z.array(z.number().int()).optional(),
            times_derived: z
              .number()
              .int()
              .min(1)
              .optional()
              .describe(
                "Rarely needed. OMIT when consolidating: retiring each source as duplicate_absorbed already moves its derivation count onto the survivor, so supplying a count here counts those sources twice. Enrichment carries the predecessor's count forward on its own.",
              ),
          }),
        ).describe("Search-backed admission decisions. Search with query_conclusions before writing."),
        agent_trace_id: z.string().describe("Agent run or trace identifier."),
        agent_model: z.string().describe("Exact model identifier making the decision."),
        session_id: z
          .string()
          .optional()
          .describe(
            "Optional: associate conclusions with a session. Omit for global conclusions.",
          ),
      },
    },
    async ({
      workspace_id,
      peer_id,
      target_peer_id,
      conclusions,
      agent_trace_id,
      agent_model,
      session_id,
    }) => {
      try {
        const peer = await ctx.clientFor(workspace_id).peer(peer_id);
        const scope = peer.conclusionsOf(target_peer_id);
        const sourceToolCallId = `mcp-create-conclusion-${crypto.randomUUID()}`;
        const params = conclusions.map((item) => ({
          content: item.content,
          sessionId: session_id,
          action: item.action,
          targetId: item.target_id,
          reasonForEntry: item.reason_for_entry,
          searchQuery: item.search_query,
          searchedConclusionIds: item.searched_conclusion_ids,
          sourceMessageIds: item.source_message_ids,
          sourceToolCallId: sourceToolCallId,
          entryOrigin: "explicit_agent" as const,
          agentTraceId: agent_trace_id,
          agentModel: agent_model,
          timesDerived: item.times_derived,
        }));
        const created = await scope.create(params);
        return textResult({
          created: created.length,
          conclusions: created.map((c) => ({
            id: c.id,
            content: c.content,
            times_derived: c.timesDerived,
          })),
        });
      } catch (e) {
        return errorResult(
          `Failed to create conclusions: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── delete_conclusion ───────────────────────────────────────────────
  server.registerTool(
    "delete_conclusion",
    {
      description: [
        "Retire a specific conclusion by ID, recording why.",
        "Use query_conclusions or list_conclusions to find the ID first.",
        "The row is kept with your reason; it stops being returned by search and stops",
        "reaching any deriving agent. When another conclusion carries the same memory,",
        "use duplicate_absorbed with absorbed_into so its derivation count moves there.",
      ].join("\n"),
      inputSchema: {
        workspace_id: workspaceIdSchema(ctx),
        peer_id: z.string().describe("The observer peer."),
        target_peer_id: z
          .string()
          .describe("The peer the conclusion is about."),
        conclusion_id: z.string().describe("The conclusion to retire."),
        category: z
          .enum([
            "duplicate_absorbed",
            "superseded",
            "contradicted",
            "misderived",
            "out_of_scope",
            "transient",
            "low_value",
          ])
          .describe(
            "duplicate_absorbed: folded into another conclusion (requires absorbed_into). superseded: was true, reality changed. contradicted: was never true. misderived: extraction defect. out_of_scope: valid but belongs elsewhere, carrier verified. transient: never durable. low_value: not worth carrying.",
          ),
        reason: z
          .string()
          .describe(
            "Specific justification, naming what carries the memory now when something does.",
          ),
        absorbed_into: z
          .string()
          .optional()
          .describe(
            "Required for duplicate_absorbed: the surviving conclusion, which inherits this one's derivation count.",
          ),
        agent_trace_id: z.string().describe("Agent run or trace identifier."),
        agent_model: z
          .string()
          .describe("Exact model identifier making the decision."),
      },
    },
    async ({
      workspace_id,
      peer_id,
      target_peer_id,
      conclusion_id,
      category,
      reason,
      absorbed_into,
      agent_trace_id,
      agent_model,
    }) => {
      try {
        const peer = await ctx.clientFor(workspace_id).peer(peer_id);
        const scope = peer.conclusionsOf(target_peer_id);
        await scope.delete(conclusion_id, {
          category,
          reason,
          absorbed_into,
          entry_origin: "explicit_agent",
          agent_trace_id,
          agent_model,
        });
        return textResult("Conclusion retired successfully");
      } catch (e) {
        return errorResult(
          `Failed to retire conclusion: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── conclusion_lineage ──────────────────────────────────────────────
  server.registerTool(
    "conclusion_lineage",
    {
      description: [
        "Get one conclusion's full ledger, whether it is live or retired.",
        "Returns how it was admitted, every prior formulation it was rewritten from,",
        "everything it absorbed with the count each brought, and why it was removed.",
        "This is the only way to read a retired conclusion; search never returns one.",
      ].join("\n"),
      inputSchema: {
        workspace_id: workspaceIdSchema(ctx),
        peer_id: z.string().describe("The observer peer."),
        target_peer_id: z
          .string()
          .describe("The peer the conclusion is about."),
        conclusion_id: z.string().describe("The conclusion to trace."),
      },
    },
    async ({ workspace_id, peer_id, target_peer_id, conclusion_id }) => {
      try {
        const peer = await ctx.clientFor(workspace_id).peer(peer_id);
        const scope = peer.conclusionsOf(target_peer_id);
        const lineage = await scope.lineage(conclusion_id);
        return textResult({
          id: lineage.id,
          content: lineage.content,
          times_derived: lineage.timesDerived,
          created_at: lineage.createdAt,
          deleted_at: lineage.deletedAt,
          admission: lineage.admission,
          removal: lineage.removal,
          admission_history: lineage.admissionHistory,
          absorbed: lineage.absorbed,
        });
      } catch (e) {
        return errorResult(
          `Failed to read conclusion lineage: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );
}

File: src/routers/conclusions.py
import logging

from fastapi import APIRouter, Body, Depends, Path, Query
from fastapi_pagination import Page
from fastapi_pagination.ext.sqlalchemy import apaginate
from sqlalchemy.ext.asyncio import AsyncSession

from src import crud, schemas
from src.dependencies import db, read_db
from src.exceptions import ResourceNotFoundException, ValidationException
from src.security import require_auth
from src.telemetry.events import EmbeddingCallPurpose
from src.utils.types import embedding_call_purpose

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/workspaces/{workspace_id}/conclusions",
    tags=["conclusions"],
    dependencies=[Depends(require_auth(workspace_name="workspace_id"))],
)


@router.post(
    "",
    response_model=list[schemas.Conclusion],
    status_code=201,
)
async def create_conclusions(
    workspace_id: str = Path(...),
    body: schemas.ConclusionBatchCreate = Body(
        ...,
        description="Batch of Conclusions to create",
    ),
    db: AsyncSession = db,
) -> list[schemas.Conclusion]:
    """
    Create one or more Conclusions.

    Conclusions are logical certainties derived from interactions between Peers. They form the basis of a Peer's Representation.
    """
    documents = await crud.create_observations(
        db,
        observations=body.conclusions,
        workspace_name=workspace_id,
    )

    logger.debug(
        "Created %d conclusions in workspace %s",
        len(documents),
        workspace_id,
    )
    return [schemas.Conclusion.model_validate(doc) for doc in documents]


@router.post(
    "/list",
    response_model=Page[schemas.Conclusion],
)
async def list_conclusions(
    workspace_id: str = Path(...),
    options: schemas.ConclusionGet | None = Body(
        None,
        description="Filtering options for the Conclusions list",
    ),
    reverse: bool | None = Query(
        False,
        description="Whether to reverse the order of results",
    ),
    include_deleted: bool = Query(
        False,
        description="Include retired Conclusions, each carrying the reason it was removed",
    ),
    db: AsyncSession = read_db,
):
    """
    List Conclusions using optional filters, ordered by recency unless `reverse` is true. Results are paginated.
    """
    filters = None
    if options and hasattr(options, "filters"):
        filters = options.filters
        if filters == {}:
            filters = None

    stmt = crud.get_documents_with_filters(
        workspace_name=workspace_id,
        filters=filters,
        reverse=reverse or False,
        include_deleted=include_deleted,
    )

    return await apaginate(db, stmt)


@router.post(
    "/query",
    response_model=list[schemas.Conclusion],
)
async def query_conclusions(
    workspace_id: str = Path(...),
    body: schemas.ConclusionQuery = Body(
        ...,
        description="Semantic search parameters for Conclusions",
    ),
    db: AsyncSession = read_db,
) -> list[schemas.Conclusion]:
    """
    Query Conclusions using semantic search. Use `top_k` to control the number of results returned.
    """
    observer = None
    observed = None
    if body.filters:
        observer = body.filters.get("observer") or body.filters.get("observer_id")
        observed = body.filters.get("observed") or body.filters.get("observed_id")

    if not observer or not observed:
        raise ValidationException(
            "observer and observed must be specified for semantic search. "
            + "Pass them inside the 'filters' object, e.g. "
            + '{"query": "...", "filters": {"observer": "alice", "observed": "bob"}}. '
            + "Both 'observer'/'observer_id' and 'observed'/'observed_id' are accepted."
        )

    with embedding_call_purpose(
        EmbeddingCallPurpose.GENERIC_DOCUMENT_SEARCH.value,
        workspace_name=workspace_id,
        parent_category="api",
    ):
        documents = await crud.query_documents(
            db,
            workspace_name=workspace_id,
            query=body.query,
            observer=observer,
            observed=observed,
            filters=body.filters,
            max_distance=body.distance,
            top_k=body.top_k,
        )
    return [schemas.Conclusion.model_validate(doc) for doc in documents]


@router.get(
    "/{conclusion_id}/lineage",
    response_model=schemas.ConclusionDetail,
)
async def get_conclusion_lineage(
    workspace_id: str = Path(...),
    conclusion_id: str = Path(...),
    db: AsyncSession = read_db,
) -> schemas.ConclusionDetail:
    """
    Get one Conclusion's full ledger: how it was admitted, every prior formulation it
    has been rewritten from, everything it absorbed, and why it was removed if it was.

    Retired Conclusions are returned here; they are never returned by search.
    """
    document = await crud.get_document(
        db,
        workspace_name=workspace_id,
        document_id=conclusion_id,
    )
    if document is None:
        raise ResourceNotFoundException("Conclusion not found")
    return schemas.ConclusionDetail.model_validate(document)


@router.delete(
    "/{conclusion_id}",
    status_code=204,
    response_model=None,
)
async def delete_conclusion(
    workspace_id: str = Path(...),
    conclusion_id: str = Path(...),
    body: schemas.ConclusionRemoval = Body(
        ...,
        description="Why this Conclusion is being retired",
    ),
    db: AsyncSession = db,
):
    """
    Retire a single Conclusion by ID, recording why.

    The row and its ledger are kept; the Conclusion stops being returned by search and
    stops reaching any deriving agent. Use `duplicate_absorbed` with `absorbed_into` when
    another Conclusion carries the memory, so its derivation count moves with it.
    """
    if body.category not in schemas.AGENT_REMOVAL_CATEGORIES:
        raise ValidationException(
            f"'{body.category}' is recorded by Honcho itself and cannot be supplied. "
            + f"Choose one of: {', '.join(sorted(schemas.AGENT_REMOVAL_CATEGORIES))}."
        )
    try:
        await crud.delete_document_by_id(
            db,
            workspace_name=workspace_id,
            document_id=conclusion_id,
            removal=body,
        )

        logger.debug("Conclusion %s retired successfully", conclusion_id)
    except ResourceNotFoundException:
        raise
    except ValueError as e:
        logger.warning(f"Failed to retire conclusion {conclusion_id}: {str(e)}")
        raise ResourceNotFoundException("Conclusion not found") from e
